domain/model/service/service_test.go

```diff
@@ -1112,3 +1122,93 @@ func (s *serviceSuite) TestDefaultModelCloudNameAndCredential(c *gc.C) {
 		Name:  "test-cred",
 	})
 }
+
+// TestWatchActivatedModels verifies that WatchActivatedModels correctly sets up a watcher
+// that emits events for activated models when the watcher receives change events.
+func (s *serviceSuite) TestWatchActivatedModels(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	ctx := context.Background()
+	svc := NewWatchableService(
+		s.mockState,
+		s.mockModelDeleter,
+		loggertesting.WrapCheckLog(c),
+		s.mockWatcherFactory,
+	)
+
+	// Set up necessary mock return values.
+	s.mockState.EXPECT().GetAllActivatedModelsUUIDQuery().Return(
+		"SELECT uuid from model WHERE activated = true",
+	)
+
+	changes := make(chan []string, 1)
+	activatedModelUUID1 := modeltesting.GenModelUUID(c)
+	activatedModelUUID2 := modeltesting.GenModelUUID(c)
+	activatedModelUUIDs := []coremodel.UUID{activatedModelUUID1, activatedModelUUID2}
+	activatedModelUUIDsStr := transform.Slice(activatedModelUUIDs, func(uuid coremodel.UUID) string {
+		return uuid.String()
+	})
+	changes <- activatedModelUUIDsStr
+	s.mockStringsWatcher.EXPECT().Changes().AnyTimes().Return(changes)
+
+	s.mockWatcherFactory.EXPECT().NewNamespaceMapperWatcher("model", changestream.Changed,
+		gomock.Any(), gomock.Any()).Return(s.mockStringsWatcher, nil)
+
+	// Verifies that the service returns a watcher with the correct model UUIDs string.
+	watcher, err := svc.WatchActivatedModels(ctx)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(<-watcher.Changes(), gc.DeepEquals, activatedModelUUIDsStr)
+}
+
+func (s *serviceSuite) createMockChangeEventsFromUUIDs(uuids ...coremodel.UUID) []changestream.ChangeEvent {
+	events := make([]changestream.ChangeEvent, len(uuids))
+	for i, uuid := range uuids {
+		event := changestreammock.NewMockChangeEvent(s.mockWatcherFactory.ctrl)
+		event.EXPECT().Changed().AnyTimes().Return(
+			uuid.String(),
+		)
+		events[i] = event
+	}
+	return events
+}
+
+// TestWatchActivatedModelsMapper verifies that the WatchActivatedModelsMapper correctly
+// filters change events to include only those associated with activated models and that
+// the subset of changes returned is maintained in the same order as they are received.
+func (s *serviceSuite) TestWatchActivatedModelsMapper(c *gc.C) {
```

Comment:
As stated above I think this needs to test the watcher function and not the internal mapper function.

Reply:
The TestWatchActivatedModels function tests the watcher function, while this is for testing the mapper because I felt that in this case, the mapper contains the core logic of the function. However, happy to remove this and just retain the TestWatchActivatedModels as suggested :)